### DISCUSSION TITLE
Native Graph-as-Text URLs (JSON I think)

### DISCUSSION THREAD

**Discussion by WNKLER** [DesModder/DesModder#400] [General]
Maybe this is common knowledge, but it might be helpful to know that desmos hosts a native text-representation of every graph.
It is fetchable via a URL.

Take the graph with the ID `qfzqejspct` as an example:
[https://www.desmos.com/calculator/qfzqejspct](url)

Accessing this URL: [https://saved-work.desmos.com/calc-states/production/qfzqejspct](url)

returns:
```json
{
	"version":9,
	"randomSeed":"feec3866eba2d1fb44f38db220329d1b",
	"graph":{
		"viewport":{
			"xmin":-9.998391734489985,
			"ymin":-6.276256152833961,
			"xmax":10.001608265510018,
			"ymax":6.273039621813929
		}
	},
	"expressions":{
		"list":[
			{
				"type":"text",
				"id":"3",
				"text":"URL Test"
			},
			{
				"type":"expression",
				"id":"1",
				"color":"#c74440",
				"latex":"y=2x^{2}",
				"labelSize":"medium"
			},
			{
				"type":"folder",
				"id":"4",
				"title":"Test Folder"
			},
			{
				"type":"text",
				"id":"10",
				"folderId":"4",
				"text":"Folder Test"
			},
			{
				"type":"expression",
				"id":"11",
				"folderId":"4",
				"color":"#2d70b3",
				"latex":"y=\\frac{1}{e^{x}}"
			},
			{
				"id":"7",
				"type":"table",
				"folderId":"4",
				"columns":[
					{
						"values":["0","1"],
						"hidden":true,
						"id":"5",
						"color":"#388c46",
						"latex":"x_{1}"
					},
					{
						"values":["1","2"],
						"id":"6",
						"color":"#6042a6",
						"latex":"y_{1}"
					}
				]
			}
		]
	}
}
```

Note: I added whitespace for readability. The raw output has no whitespace.

You can similarly fetch a thumbnail of the graph from here:
[https://saved-work.desmos.com/calc_thumbs/production/qfzqejspct.png](url)

![qfzqejspct](https://user-images.githubusercontent.com/29347893/181350456-46f577ee-d7e1-43c8-a8b9-2ec8e07a5703.png)

URL Syntax:
`https://saved-work.desmos.com/calc-states/production/<graph_ID>`
`https://saved-work.desmos.com/calc_thumbs/production/<graph_ID>.png`

**Comment by jared-hughes** [comment]
Thanks for sharing! Do you have any corresponding feature request for DesModder or is this just a discussion/fyi?

Btw this can also be achieved through an `Accept` header on the original link, and it gives a little more information

```js
const stream = await fetch("https://www.desmos.com/calculator/qfzqejspct", {
  headers: {
    Accept: "application/json",
  },
});
const data = await stream.json();
```
The `data` variable then holds:
```json
{
    "hash": "qfzqejspct",
    "parent_hash": null,
    "thumbUrl": "https://saved-work.desmos.com/calc_thumbs/production/qfzqejspct.png",
    "stateUrl": "https://saved-work.desmos.com/calc-states/production/qfzqejspct",
    "title": null,
    "access": "link",
    "created": "Wed, 27 Jul 2022 18:15:16 GMT",
    "state": {
        "version": 9,
        "randomSeed": "feec3866eba2d1fb44f38db220329d1b",
        "graph": {
            "viewport": {
                "xmin": -9.998391734489985,
                "ymin": -6.276256152833961,
                "xmax": 10.001608265510018,
                "ymax": 6.273039621813929
            }
        },
        "expressions": {
            "list": [ (etc. same as your original post) ]
        }
    }
}
```

**Comment by WNKLER** [comment]
Oh, even cooler!

I created the issue just as and FYI.

There ARE two desmos features on my Wishlist, but I believe they're out of reach for a tool like DesModder.

1. Graph URLs that don't change after every update (anyone with a link to my work should be taken to the latest version)
2. Keywords for the Viewport's xmin/ymin/xmax/ymax that can be used in expressions

Thanks Jared,
You're doing some fantastic work on here!
-Ben

**Reply by jared-hughes** [reply]
Thanks for voicing those suggestions. I also want these features :), but I agree they're out of scope

1. would require a separate database somewhere, which is an entirely different project from a browser extension
2. violates one of DesModder's goals, which is that any graph created with DesModder could be used by someone without any extensions. Really hoping Desmos adds it, though they have stated there's some technical challenges with it related to updating the variables while zooming/panning.

**Comment by jared-hughes** [comment]
To encourage more discussion, I've enabled the discussions feature on Github. Going to migrate this issue to a discussion post.
